File: backend/app/core/pagespeed.py
"""
Optional Google PageSpeed Insights integration (spec 12, 35).

Disabled unless the user supplies their own API key in Settings. The key is
held server-side only and never returned to the frontend unmasked. If the
API is unavailable, the result carries `measured: False` and the audit
reports no performance score rather than inventing one.
"""

from __future__ import annotations

import asyncio
from typing import Any, Dict, Optional

import httpx

ENDPOINT = "https://www.googleapis.com/pagespeedonline/v5/runPagespeed"

_semaphore = asyncio.Semaphore(2)  # respect the API's rate limits


async def measure(
    url: str,
    api_key: str,
    strategy: str = "mobile",
    timeout_s: float = 60.0,
) -> Dict[str, Any]:
    out: Dict[str, Any] = {
        "measured": False,
        "strategy": strategy,
        "url": url,
        "source": "Google PageSpeed Insights API",
    }
    if not api_key:
        out["error"] = "not_configured"
        out["error_message"] = "No PageSpeed API key is configured."
        return out

    params = {
        "url": url,
        "key": api_key,
        "strategy": strategy,
        "category": "performance",
    }

    try:
        async with _semaphore:
            async with httpx.AsyncClient(timeout=timeout_s) as client:
                resp = await client.get(ENDPOINT, params=params)
    except httpx.TimeoutException:
        out["error"] = "timeout"
        out["error_message"] = f"PageSpeed did not respond within {timeout_s:.0f}s."
        return out
    except httpx.HTTPError as exc:
        out["error"] = "http_error"
        out["error_message"] = f"PageSpeed request failed: {exc}"
        return out

    if resp.status_code != 200:
        out["error"] = f"http_{resp.status_code}"
        try:
            detail = resp.json().get("error", {}).get("message", "")
        except Exception:
            detail = resp.text[:300]
        out["error_message"] = f"PageSpeed returned HTTP {resp.status_code}: {detail}"
        return out

    try:
        data = resp.json()
    except Exception as exc:
        out["error"] = "parse_error"
        out["error_message"] = f"Could not parse the PageSpeed response: {exc}"
        return out

    lighthouse = data.get("lighthouseResult") or {}
    categories = lighthouse.get("categories") or {}
    audits = lighthouse.get("audits") or {}

    perf = (categories.get("performance") or {}).get("score")
    if perf is None:
        out["error"] = "no_score"
        out["error_message"] = "PageSpeed returned no performance score for this URL."
        return out

    out["measured"] = True
    out["performance_score"] = round(perf * 100)
    out["fetched_url"] = lighthouse.get("finalUrl") or url
    out["lighthouse_version"] = lighthouse.get("lighthouseVersion", "")

    def num(key: str) -> Optional[float]:
        a = audits.get(key) or {}
        v = a.get("numericValue")
        return round(float(v), 3) if isinstance(v, (int, float)) else None

    lcp = num("largest-contentful-paint")
    fcp = num("first-contentful-paint")
    tbt = num("total-blocking-time")
    cls = num("cumulative-layout-shift")
    si = num("speed-index")

    out["lcp_s"] = round(lcp / 1000, 2) if lcp is not None else None
    out["fcp_s"] = round(fcp / 1000, 2) if fcp is not None else None
    out["tbt_ms"] = round(tbt) if tbt is not None else None
    out["cls"] = round(cls, 3) if cls is not None else None
    out["speed_index_s"] = round(si / 1000, 2) if si is not None else None

    opportunities = []
    for key, audit in audits.items():
        details = audit.get("details") or {}
        if details.get("type") != "opportunity":
            continue
        savings = details.get("overallSavingsMs")
        if isinstance(savings, (int, float)) and savings >= 250:
            opportunities.append({
                "id": key,
                "title": audit.get("title", ""),
                "savings_ms": round(savings),
            })
    opportunities.sort(key=lambda o: -o["savings_ms"])
    out["opportunities"] = opportunities[:6]

    # Field data (CrUX), when Google has enough real-user traffic for the site.
    loading = data.get("loadingExperience") or {}
    if loading.get("overall_category"):
        out["field_data_category"] = loading["overall_category"]

    return out


async def check_availability(api_key: str) -> Dict[str, Any]:
    """Used by the System Health page."""
    if not api_key:
        return {"status": "disabled", "detail": "No API key configured (optional integration)."}
    result = await measure("https://example.com", api_key, "mobile", timeout_s=30.0)
    if result.get("measured"):
        return {"status": "healthy", "detail": "PageSpeed API responded successfully."}
    return {
        "status": "error",
        "detail": result.get("error_message", "PageSpeed API did not return a score."),
    }

File: cloudflare/src/core/pagespeed.ts
// Optional Google PageSpeed Insights integration. Ported from
// backend/app/core/pagespeed.py. Off unless the user supplies their own key
// in Settings.
const ENDPOINT = "https://www.googleapis.com/pagespeedonline/v5/runPagespeed";

export interface PageSpeedResult {
  measured: boolean;
  strategy: string;
  url: string;
  source: string;
  error?: string;
  error_message?: string;
  performance_score?: number;
  fetched_url?: string;
  lcp_s?: number | null;
  fcp_s?: number | null;
  tbt_ms?: number | null;
  cls?: number | null;
  speed_index_s?: number | null;
  opportunities?: { id: string; title: string; savings_ms: number }[];
  field_data_category?: string;
}

export async function measure(url: string, apiKey: string, strategy: "mobile" | "desktop" = "mobile", timeoutS = 55): Promise<PageSpeedResult> {
  const out: PageSpeedResult = { measured: false, strategy, url, source: "Google PageSpeed Insights API" };
  if (!apiKey) {
    out.error = "not_configured";
    out.error_message = "No PageSpeed API key is configured.";
    return out;
  }

  const params = new URLSearchParams({ url, key: apiKey, strategy, category: "performance" });
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutS * 1000);
  let resp: Response;
  try {
    resp = await fetch(`${ENDPOINT}?${params.toString()}`, { signal: controller.signal });
  } catch (exc: any) {
    out.error = exc?.name === "AbortError" ? "timeout" : "http_error";
    out.error_message = exc?.name === "AbortError" ? `PageSpeed did not respond within ${timeoutS}s.` : `PageSpeed request failed: ${exc?.message || exc}`;
    return out;
  } finally {
    clearTimeout(timer);
  }

  if (resp.status !== 200) {
    out.error = `http_${resp.status}`;
    let detail = "";
    try {
      detail = ((await resp.json()) as any)?.error?.message || "";
    } catch {
      detail = (await resp.text()).slice(0, 300);
    }
    out.error_message = `PageSpeed returned HTTP ${resp.status}: ${detail}`;
    return out;
  }

  let data: any;
  try {
    data = await resp.json();
  } catch (exc: any) {
    out.error = "parse_error";
    out.error_message = `Could not parse the PageSpeed response: ${exc?.message || exc}`;
    return out;
  }

  const lighthouse = data.lighthouseResult || {};
  const categories = lighthouse.categories || {};
  const audits = lighthouse.audits || {};
  const perf = categories.performance?.score;
  if (perf === undefined || perf === null) {
    out.error = "no_score";
    out.error_message = "PageSpeed returned no performance score for this URL.";
    return out;
  }

  out.measured = true;
  out.performance_score = Math.round(perf * 100);
  out.fetched_url = lighthouse.finalUrl || url;

  const num = (key: string): number | null => {
    const v = audits[key]?.numericValue;
    return typeof v === "number" ? Math.round(v * 1000) / 1000 : null;
  };
  const lcp = num("largest-contentful-paint");
  const fcp = num("first-contentful-paint");
  const tbt = num("total-blocking-time");
  const cls = num("cumulative-layout-shift");
  const si = num("speed-index");

  out.lcp_s = lcp !== null ? Math.round((lcp / 1000) * 100) / 100 : null;
  out.fcp_s = fcp !== null ? Math.round((fcp / 1000) * 100) / 100 : null;
  out.tbt_ms = tbt !== null ? Math.round(tbt) : null;
  out.cls = cls !== null ? Math.round(cls * 1000) / 1000 : null;
  out.speed_index_s = si !== null ? Math.round((si / 1000) * 100) / 100 : null;

  const opportunities: { id: string; title: string; savings_ms: number }[] = [];
  for (const [key, audit] of Object.entries<any>(audits)) {
    const details = audit.details || {};
    if (details.type !== "opportunity") continue;
    const savings = details.overallSavingsMs;
    if (typeof savings === "number" && savings >= 250) {
      opportunities.push({ id: key, title: audit.title || "", savings_ms: Math.round(savings) });
    }
  }
  opportunities.sort((a, b) => b.savings_ms - a.savings_ms);
  out.opportunities = opportunities.slice(0, 6);

  const loading = data.loadingExperience || {};
  if (loading.overall_category) out.field_data_category = loading.overall_category;

  return out;
}

export async function checkAvailability(apiKey: string): Promise<{ status: string; detail: string }> {
  if (!apiKey) return { status: "disabled", detail: "No API key configured (optional integration)." };
  const result = await measure("https://example.com", apiKey, "mobile", 30);
  if (result.measured) return { status: "healthy", detail: "PageSpeed API responded successfully." };
  return { status: "error", detail: result.error_message || "PageSpeed API did not return a score." };
}
